x-pack/packages/ml/aiops_log_rate_analysis/queries/fetch_field_candidates.ts

```diff
@@ -0,0 +1,130 @@
+}
+
+export interface FetchFieldCandidatesParams {
+  esClient: ElasticsearchClient;
+  abortSignal?: AbortSignal;
+  arguments: AiopsLogRateAnalysisSchema & FetchFieldCandidatesParamsArguments;
+}
+
+export interface FetchFieldCandidatesResponse {
+  isECS: boolean;
+  keywordFieldCandidates: string[];
+  selectedKeywordFieldCandidates: string[];
+  textFieldCandidates: string[];
+  selectedTextFieldCandidates: string[];
+}
+
+export const fetchFieldCandidates = async ({
+  esClient,
+  abortSignal,
+  arguments: args,
+}: FetchFieldCandidatesParams): Promise<FetchFieldCandidatesResponse> => {
+  const { textFieldCandidatesOverrides = [], ...params } = args;
+
+  // Get all supported fields
+  const respMapping = await esClient.fieldCaps(
+    {
+      fields: '*',
+      filters: '-metadata,-parent',
+      include_empty_fields: false,
+      index: params.index,
+      index_filter: {
+        range: {
+          [params.timeFieldName]: {
+            gte: params.deviationMin,
+            lte: params.deviationMax,
+          },
+        },
+      },
+      types: [...SUPPORTED_ES_FIELD_TYPES, ...SUPPORTED_ES_FIELD_TYPES_TEXT],
+    },
+    { signal: abortSignal, maxRetries: 0 }
+  );
+
+  const allFieldNames: string[] = [];
+
+  const acceptableFields: Set<string> = new Set();
+  const acceptableTextFields: Set<string> = new Set();
+
+  Object.entries(respMapping.fields).forEach(([key, value]) => {
+    const fieldTypes = Object.keys(value) as ES_FIELD_TYPES[];
+    const isSupportedType = fieldTypes.some((type) => SUPPORTED_ES_FIELD_TYPES.includes(type));
+    const isAggregatable = fieldTypes.some((type) => value[type].aggregatable);
+    const isTextField = fieldTypes.some((type) => SUPPORTED_ES_FIELD_TYPES_TEXT.includes(type));
+
+    // Check if fieldName is something we can aggregate on
+    if (isSupportedType && isAggregatable) {
+      acceptableFields.add(key);
+    }
+
+    if (isTextField && TEXT_FIELD_SAFE_LIST.includes(key)) {
+      acceptableTextFields.add(key);
+    }
+
+    allFieldNames.push(key);
+  });
+
+  const textFieldCandidatesOverridesWithKeywordPostfix = textFieldCandidatesOverrides.map(
+    (d) => `${d}.keyword`
+  );
+
+  const keywordFieldCandidates: string[] = [...acceptableFields].filter(
+    (field) => !textFieldCandidatesOverridesWithKeywordPostfix.includes(field)
+  );
+  const textFieldCandidates: string[] = [...acceptableTextFields].filter((field) => {
+    const fieldName = field.replace(new RegExp(/\.text$/), '');
+    return (
+      (!keywordFieldCandidates.includes(fieldName) &&
+        !keywordFieldCandidates.includes(`${fieldName}.keyword`)) ||
+      textFieldCandidatesOverrides.includes(field)
+    );
+  });
+
+  const isECS = containsECSIdentifierField(keywordFieldCandidates);
+
+  return {
+    isECS,
+    // all keyword field candidates
+    keywordFieldCandidates: keywordFieldCandidates.sort(),
+    // preselection:
+    // - if we identify an ECS schema, filter by custom ECS whitelist
```

Comment:
Tiny nit: change 'white' to 'safe' as was changed in the other file

Reply:
Updated in [4606cdc](https://github.com/elastic/kibana/pull/188913/commits/4606cdc28ed9a614f8e79495b51a2b0592d78899).